File: prototypes/prototype.py
def maximalSquare(matrix):
    #todo: save full, empty, obstacle chars in vars.
    #todo: check maps conditions
    if not matrix or not matrix[0]:
        return matrix

    rows, cols = len(matrix), len(matrix[0])
    # DP table to store the size of the largest square submatrix ending at each position
    dp = [[0] * cols for _ in range(rows)]
    atwal_9otr = 0  # Maximum side length of the square found
    max_bottom_right = (0, 0)  # Bottom-right corner of the largest square

    # Iterate over the matrix to fill the DP table
    for y in range(rows):
        for x in range(cols):
            if matrix[y][x] == '1':
                # For the first row or column, the square can only be of size 1 if matrix[i][j] is '1'
                if y == 0 or x == 0:
                    dp[y][x] = 1
                else:
                    # DP relation: min of top, left, and top-left diagonal + 1
                    dp[y][x] = min(dp[y-1][x], dp[y][x-1], dp[y-1][x-1]) + 1

                # Update max side length and the bottom-right corner of the largest square
                if dp[y][x] > atwal_9otr:
                    atwal_9otr = dp[y][x]
                    max_bottom_right = (y, x)

    # Retrieve the bottom-right corner and side length of the largest square
    y_bot_r, x_bot_r = max_bottom_right

    # Replace the largest square with 'X'
    for y in range(y_bot_r - atwal_9otr + 1, y_bot_r + 1):
        for x in range(x_bot_r - atwal_9otr + 1, x_bot_r + 1):
            matrix[y][x] = 'X'

    return matrix

# Example usage
matrix = [
    ["1","1","1","1","1","1"],
    ["1","0","1","0","1","1"],
    ["1","1","1","1","1","0"],
    ["1","1","0","1","1","1"],
    ["1","1","1","1","1","1"],
    ["1","1","1","1","1","1"]
]

result = maximalSquare(matrix)

for row in result:
    print(" ".join(row))

File: prototypes/prototype.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void bsq(char **matrix, int rows, int cols)
{
	// ! put this in necessary checks
	if (rows == 0 || cols == 0)
		return;

	// Allocate memory for the DP table
	int **dp = (int **)malloc(rows * sizeof(int *));
	for (int i = 0; i < rows; i++)
	{
		dp[i] = (int *)malloc(cols * sizeof(int));
		memset(dp[i], 0, cols * sizeof(int)); // Initialize dp array to 0
	}

	int atwal_9otr = 0; // Maximum side length of the square found
	int y_bot_r = 0, x_bot_r = 0;

	// Fill the DP table
	for (int y = 0; y < rows; y++)
	{
		for (int x = 0; x < cols; x++)
		{
			if (matrix[y][x] == '1')
			{
				if (y == 0 || x == 0)
					dp[y][x] = 1;
				else
				{
					dp[y][x] = (dp[y - 1][x] < dp[y][x - 1] ? dp[y - 1][x] : dp[y][x - 1]);
					dp[y][x] = (dp[y][x] < dp[y - 1][x - 1] ? dp[y][x] : dp[y - 1][x - 1]) + 1;
				}

				if (dp[y][x] > atwal_9otr)
				{
					atwal_9otr = dp[y][x];
					y_bot_r = y;
					x_bot_r = x;
				}
			}
		}
	}

	// Replace the largest square with 'X'
	for (int y = y_bot_r - atwal_9otr + 1; y <= y_bot_r; y++)
	{
		for (int x = x_bot_r - atwal_9otr + 1; x <= x_bot_r; x++)
		{
			matrix[y][x] = 'X';  //! replace this with full_char
		}
	}

	// Free memory
	for (int r = 0; r < rows; r++)
		free(dp[r]);
	free(dp);
}

// Helper function to print the matrix
void printMatrix(char **matrix, int rows, int cols)
{
	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < cols; j++)
		{
			printf("%c", matrix[i][j]);
		}
		printf("\n");
	}
}

int main()
{
	// Define the matrix
	int rows = 6, cols = 6;
	char *matrix[] = {
		"111111",
		"101011",
		"111111",
		"111111",
		"111101",
		"011111"};

	// Allocate memory for the matrix
	char **matrix_ptr = (char **)malloc(rows * sizeof(char *));
	for (int i = 0; i < rows; i++)
	{
		matrix_ptr[i] = (char *)malloc((cols + 1) * sizeof(char)); // +1 for null terminator
		strncpy(matrix_ptr[i], matrix[i], cols);
		matrix_ptr[i][cols] = '\0'; // Null-terminate each row
	}

	printf("Original Matrix:\n");
	printMatrix(matrix_ptr, rows, cols);

	// Find and update the maximal square
	bsq(matrix_ptr, rows, cols);

	printf("\nMatrix After Maximal Square Update:\n");
	printMatrix(matrix_ptr, rows, cols);

	// Free allocated memory
	for (int i = 0; i < rows; i++)
	{
		free(matrix_ptr[i]);
	}
	free(matrix_ptr);

	return 0;
}
